File: backend/agent/tools/linkedin.py
"""
Phantom — LinkedIn Signals Tool (via Web Unlocker)

Scrapes LinkedIn company pages in real-time using Bright Data Web Unlocker
to extract headcount and growth signals.

Same rationale as indeed.py — uses Web Unlocker (synchronous/real-time)
rather than the Datasets batch API, which is async and incompatible with
live streaming responses.

Input: company name
Output: LinkedInSignals with headcount and headcount_delta_pct
"""

import os
import re
import logging
import httpx
from urllib.parse import quote_plus
from typing import Optional

from models.schemas import LinkedInSignals

logger = logging.getLogger(__name__)

UNLOCKER_API_URL = "https://api.brightdata.com/request"


async def fetch_linkedin_signals(
    query: str,
    company: Optional[str] = None,
    **kwargs,
) -> Optional[LinkedInSignals]:
    """
    Scrape LinkedIn company page via Bright Data Web Unlocker.

    Retrieves the company's LinkedIn profile page and parses
    the employee count. Headcount delta is inferred from context
    clues (growth mentions) when a second data point isn't available.

    Args:
        query: Job search query (used if company not specified)
        company: Company name — used to build the LinkedIn search URL

    Returns:
        LinkedInSignals with headcount data, or None if the request fails.
    """
    api_key = os.getenv("BRIGHT_DATA_API_KEY")
    zone = os.getenv("BRIGHT_DATA_WEB_UNLOCKER_ZONE", "web_unlocker")

    if not api_key:
        logger.error("BRIGHT_DATA_API_KEY not set")
        return None

    search_target = company or query

    # Use LinkedIn's people search to find the company, then get its page.
    # Searching via LinkedIn search is more reliable than guessing the slug.
    encoded_company = quote_plus(search_target)
    linkedin_url = f"https://www.linkedin.com/company/{_slugify(search_target)}/about/"

    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {api_key}",
    }

    payload = {
        "zone": zone,
        "url": linkedin_url,
        "format": "json",          # Required by Bright Data
        "data_format": "markdown", # Return page as markdown for parsing
    }

    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.post(
                UNLOCKER_API_URL,
                json=payload,
                headers=headers,
            )
            response.raise_for_status()

            try:
                data = response.json()
                content = data.get("content", data.get("text", str(data)))
            except Exception:
                content = response.text

        signals = _parse_linkedin_markdown(content, search_target)

        if signals:
            logger.info(
                f"LinkedIn Scraper: headcount={signals.headcount}, "
                f"delta={signals.headcount_delta_pct}% for: '{search_target}'"
            )
        else:
            logger.warning(f"LinkedIn Scraper: no parseable signals for '{search_target}'")

        return signals

    except httpx.HTTPStatusError as e:
        logger.error(
            f"LinkedIn Scraper HTTP error: {e.response.status_code} — {e.response.text[:300]}"
        )
        return None
    except httpx.RequestError as e:
        logger.error(f"LinkedIn Scraper request error: {e}")
        return None
    except Exception as e:
        logger.error(f"LinkedIn Scraper unexpected error: {e}")
        return None


def _parse_linkedin_markdown(content: str, company: str) -> Optional[LinkedInSignals]:
    """
    Parse LinkedIn company page markdown for headcount signals.

    LinkedIn shows employee count as "X,XXX employees" or "X followers • X employees"
    on company About pages. Growth clues ("growing", "hiring") are used to estimate
    headcount_delta_pct when exact delta data is unavailable.

    Args:
        content: Page content as markdown string
        company: Company name for context

    Returns:
        LinkedInSignals or None if no headcount data found
    """
    if not content or len(content) < 50:
        return None

    content_lower = content.lower()
    headcount: Optional[int] = None
    headcount_delta_pct: Optional[float] = None

    # Match "X employees", "X,XXX employees", "X followers • X employees"
    # LinkedIn formats: "1,234 employees", "10,001+ employees", "51-200 employees"
    patterns = [
        r"([\d,]+)\s+employee",        # "1,234 employees"
        r"([\d,]+)\+?\s+employee",     # "10,001+ employees"
        r"(\d+)\s*-\s*(\d+)\s+employee",  # "51-200 employees" (range — take midpoint)
    ]

    for pattern in patterns:
        match = re.search(pattern, content_lower)
        if match:
            if len(match.groups()) == 2:
                # Range like "51-200" — take midpoint
                low = int(match.group(1).replace(",", ""))
                high = int(match.group(2).replace(",", ""))
                headcount = (low + high) // 2
            else:
                headcount = int(match.group(1).replace(",", ""))
            break

    # ─── Infer headcount delta from text signals ─────────────────
    # We can't get exact delta without two data points over time,
    # but we can use strong text signals as directional indicators.
    if headcount is not None:
        if any(kw in content_lower for kw in ["actively hiring", "growing fast", "rapid growth", "we're growing"]):
            headcount_delta_pct = 8.0  # Positive growth signal
        elif any(kw in content_lower for kw in ["layoff", "downsizing", "restructuring", "freeze"]):
            headcount_delta_pct = -5.0  # Contraction signal
        else:
            # No strong directional signal — treat as flat (most conservative for scoring)
            headcount_delta_pct = None

    if headcount is None:
        return None

    return LinkedInSignals(
        headcount=headcount,
        headcount_delta_pct=headcount_delta_pct,
        recent_posts=None,
        source="LinkedIn Scraper",
    )


def _slugify(company_name: str) -> str:
    """
    Convert company name to a LinkedIn URL slug.

    LinkedIn slugs are lowercase, hyphen-separated, no special chars.
    Examples:
        "Stripe" → "stripe"
        "Goldman Sachs" → "goldman-sachs"
        "JPMorgan Chase & Co." → "jpmorgan-chase-co"
    """
    slug = company_name.lower().strip()
    slug = re.sub(r"[&]+", "", slug)
    slug = re.sub(r"[^a-z0-9\s-]", "", slug)
    slug = re.sub(r"\s+", "-", slug)
    slug = re.sub(r"-+", "-", slug)
    return slug.strip("-")

File: src/lib/agents/linkedin.ts
/**
 * LinkedIn Agent — Fetches company headcount and growth signals via Bright Data.
 *
 * Step 1: Uses SERP API to find the LinkedIn company page URL (Google search)
 * Step 2: Uses Web Unlocker API to fetch the LinkedIn page as raw HTML
 *
 * Extracts:
 * - headcount: Current employee count
 * - headcount_delta_pct: Growth trajectory (positive = growing, negative = shrinking)
 *
 * @module agents/linkedin
 */
import { getEnv } from "@/lib/config/env";
import { BRIGHT_DATA_API_URL } from "@/lib/data";
import type { LinkedInSignals } from "@/lib/types";

export async function fetchLinkedInSignals(query: string, company?: string): Promise<{ data: LinkedInSignals; error?: string }> {
  const env = getEnv();
  const searchTarget = company || query;

  // Step 1: Find LinkedIn company URL via Google SERP API
  const linkedinUrl = await resolveLinkedInUrl(searchTarget, env.BRIGHT_DATA_API_KEY, env.BRIGHT_DATA_SERP_ZONE)
    ?? `https://www.linkedin.com/company/${slugify(searchTarget)}/about/`;

  // Step 2: Fetch the actual LinkedIn page via Web Unlocker (handles anti-bot)
  const headers = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${env.BRIGHT_DATA_API_KEY}`,
  };

  // Use Web Unlocker zone (REST API) — NOT Browser API zone
  const payload = {
    zone: env.BRIGHT_DATA_WEB_UNLOCKER_ZONE,
    url: linkedinUrl,
    format: "raw",  // Required by Bright Data docs — returns raw HTML
  };

  try {
    const response = await fetch(BRIGHT_DATA_API_URL, {
      method: "POST",
      headers,
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(30_000),
    });

    if (!response.ok) {
      const errMsg = `LinkedIn: HTTP ${response.status}`;
      console.error(errMsg);
      return {
        data: {
          headcount: null,
          headcount_delta_pct: null,
          recent_posts: null,
          source: "LinkedIn Scraper",
        },
        error: errMsg,
      };
    }

    // Web Unlocker returns raw HTML
    const content = await response.text();

    const signals = parseLinkedInHtml(content, searchTarget);

    if (signals) {
      console.log(`LinkedIn: headcount=${signals.headcount}, delta=${signals.headcount_delta_pct}% for: "${searchTarget}"`);
      return { data: signals };
    } else {
      console.warn(`LinkedIn: no parseable signals for "${searchTarget}"`);
      return {
        data: {
          headcount: null,
          headcount_delta_pct: null,
          recent_posts: null,
          source: "LinkedIn Scraper",
        },
        error: `LinkedIn: no parseable data for ${searchTarget}`,
      };
    }
  } catch (err) {
    const errMsg = `LinkedIn: network error: ${String(err).slice(0, 100)}`;
    console.error(errMsg);
    return {
      data: {
        headcount: null,
        headcount_delta_pct: null,
        recent_posts: null,
        source: "LinkedIn Scraper",
      },
      error: errMsg,
    };
  }
}

/**
 * Resolves the actual LinkedIn company URL by searching Google via SERP API.
 * Returns the first linkedin.com/company/ URL found, or null.
 */
async function resolveLinkedInUrl(company: string, apiKey: string, zone: string): Promise<string | null> {
  const encodedQuery = encodeURIComponent(`${company} site:linkedin.com/company`);
  // brd_json=1 required for parsed JSON output from SERP API
  const searchUrl = `https://www.google.com/search?q=${encodedQuery}&brd_json=1&hl=en&num=3`;

  try {
    const response = await fetch(BRIGHT_DATA_API_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify({ zone, url: searchUrl, format: "raw" }),
      signal: AbortSignal.timeout(15_000),
    });

    if (!response.ok) return null;

    const data = (await response.json()) as Record<string, unknown>;

    // Handle brd_json=1 (results[]), parsed_light (organic[]), and full parsed (navigation[])
    let organic: Array<Record<string, unknown>> = [];
    if (Array.isArray(data.organic)) {
      organic = data.organic as Array<Record<string, unknown>>;
    } else if (Array.isArray(data.results)) {
      organic = (data.results as Array<Record<string, unknown>>).filter(
        (r) => (r.type as string) === "organic"
      );
    } else if (Array.isArray(data.navigation)) {
      organic = data.navigation as Array<Record<string, unknown>>;
    } else {
      console.warn(`[linkedin.ts] SERP (resolveLinkedInUrl): unexpected response shape. Keys: [${Object.keys(data).join(", ")}]`);
      console.warn(`[linkedin.ts] SERP (resolveLinkedInUrl): first 200 chars: ${JSON.stringify(data).slice(0, 200)}`);
    }

    for (const item of organic) {
      const url = (item.link ?? item.url ?? "") as string;
      if (/linkedin\.com\/company\/[a-z0-9-]+/i.test(url)) {
        const match = url.match(/(https?:\/\/[^/]*linkedin\.com\/company\/[a-z0-9-]+)/i);
        if (match) {
          const resolved = `${match[1]}/about/`;
          console.log(`LinkedIn: resolved "${company}" → ${resolved}`);
          return resolved;
        }
      }
    }
  } catch (err) {
    console.warn(`LinkedIn URL resolution failed for "${company}": ${err}`);
  }

  return null;
}

function parseLinkedInHtml(content: string, company: string): LinkedInSignals | null {
  if (!content || content.length < 50) return null;

  const lower = content.toLowerCase();
  let headcount: number | null = null;

  const patterns = [
    /([\d,]+)\s+employee/,
    /([\d,]+)\+?\s+employee/,
    /(\d+)\s*-\s*(\d+)\s+employee/,
  ];

  for (const pattern of patterns) {
    const match = lower.match(pattern);
    if (match) {
      if (match[2] !== undefined) {
        const low = parseInt(match[1].replace(/,/g, ""), 10);
        const high = parseInt(match[2].replace(/,/g, ""), 10);
        headcount = Math.floor((low + high) / 2);
      } else {
        headcount = parseInt(match[1].replace(/,/g, ""), 10);
      }
      break;
    }
  }

  let headcountDeltaPct: number | null = null;

  if (headcount != null) {
    if (["actively hiring", "growing fast", "rapid growth", "we're growing"].some((kw) => lower.includes(kw))) {
      headcountDeltaPct = 8.0;
    } else if (["layoff", "downsizing", "restructuring", "freeze"].some((kw) => lower.includes(kw))) {
      headcountDeltaPct = -5.0;
    } else {
      headcountDeltaPct = null;
    }
  }

  if (headcount === null) return null;

  return {
    headcount,
    headcount_delta_pct: headcountDeltaPct,
    recent_posts: null,
    source: "LinkedIn Scraper",
  };
}

function slugify(name: string): string {
  return name
    .toLowerCase()
    .trim()
    .replace(/&+/g, "")
    .replace(/[^a-z0-9\s-]/g, "")
    .replace(/\s+/g, "-")
    .replace(/-+/g, "-")
    .replace(/^-|-$/g, "");
}
